### DISCUSSION TITLE
Home Assistant Integration - Hours Spent - Need Help

### DISCUSSION THREAD

**Discussion by burninmoney** [tagyoureit/nodejs-poolController#670] [General]
Migrating from SmartThings to Home Assistant.  This is the last device at home to move... ugh... 

Equipment and software details:

- Using Virtual Controller, and RS485 directly to Pentair Pump
- Running latest version of Home Assistant 2022.12.8, with latest HACS.
- Running nodejs-poolController 6.1.0 and nodejs-poolController-webclient (don't want to go through upgrade on Raspberry Pi unless required)

I can't get the Home Assistant integration to work like I want.  Installed https://github.com/Crewski/njsPC-HA on Home Assistant.  Pointed to server:4200.  Works, but only creates pool on/off sensor and does not create any more entities such as for pool properties like RPM.  Logged issue https://github.com/Crewski/njsPC-HA/issues/9

Thought I would venture a try at using MQTT instead.  New to MQTT.  Poured over all primary sources of information.  Edited pool controller settings in config.json for MQTT.  I've confirmed in Home Assistant Mosquito MQTT Broker via a listen "#" that it is receiving all messages with the data I want, but I can't get it to auto-discover entities to add to Home Assistant, and I've failed miserably at any edits to configuration.yaml.  Also, tried to install https://github.com/crsherman/nodejs-poolController-mqtt , but I get stuck at Step 4.  Directory doesn't exist.

Failed to get njsPC-HA API call to work, and I can't get MQTT properly configured.  In both instances, entities are either not created or discovered.

At a loss -- hit so many roadblocks and need someone's help to dig out!  Thanks!

**Comment by rstrouse** [comment]
You really need to upgrade your pool controller.  7.x added the much needed functionality to be able to control many more things using the Nixie controller.  njsPC and dashPanel will allow you to spin up your controller from a user interface without all the back-end editing.

See this wiki to get started.
https://github.com/tagyoureit/nodejs-poolController/wiki/Quick-Start-Guide

And this one to manage your configuration.
https://github.com/tagyoureit/nodejs-poolController/wiki/Nixie-Controller-Setup

Then the Home Assistant integration will just work.

**Reply by burninmoney** [reply]
Thanks @rstrouse.  Upgraded today.  All working well.  If anyone else runs across this thread with a similar issue, just upgrade.  much easier.  I'm amazed by the changes in 7.7 vs 6.1.

**Reply by rstrouse** [reply]
Yep we have been busy.  You can now control all of your equipment from njsPC.  If you have a heater throw a couple of temp sensors on it and it will even manage the setpoints.

**Comment by Fumado** [comment]
I'm having some issues with Home Assistant.  My NJSPC is 7.7.0.  The problem is that the homeassistant.json file I have in the web/bindings directory is adding a space between single and body in the topics.  I.E. "state_topic: `${ctx.vars.mqttTopic}/state/temps/air`,".  I removed the {$variable} and hard-coded "nixie-singlebody" and it works, I'm getting the air temperature and the pump status.  However, I'm not getting the heater status and water temperature, maybe because it's in a different file.

Perhaps someone can point me to a correct homeassistant.json file and the file that includes the heater Info.

This is partial debug info for the pool heater:

Topic: homeassistant/climate/nixie-singlebody/nixie-singlebody-1-pool/config
Payload
name: Pool Heater
unique_id: poolController-body-1-pool
action_topic: **nixie-single body**/state/temps/bodies/1/pool/heatStatus
action_template: '{{ ''off'' if value_json.heatStatus.val==0 else ''heating'' }}'
device:
  name: Pool Controller - njsPC
  identifiers:
    - nixie-single body
  suggested_area: Pool
modes:
  - 'off'
  - heat
mode_command_template: '{"id": 1, "mode": {{0 if value==''off'' else 1}}}'
mode_command_topic: nixie-single body/state/body/heatmode
mode_state_template: '{{ ''off'' if value_json.heatMode.val==0 else ''heat'' }}'
mode_state_topic: nixie-single body/state/temps/bodies/1/pool/heatMode
max_temp: '104'
min_temp: '65'
...

**Reply by rstrouse** [reply]
If you are using njsPC-HA from @crewski then it does not require MQTT or any of the HASS interfaces to be enabled on njsPC.  It uses the socket interface.  This has many fewer moving parts and it detects all of the stuff on the pool.

All that being said are you running MQTT explorer?  Can you see these variables in the topic list?

**Reply by Fumado** [reply]
Thanks, I'll check out njsPC-HA.
<img width="833" alt="Screen Shot 2022-12-31 at 2 01 01 PM" src="https://user-images.githubusercontent.com/16650838/210156436-888a9beb-47c5-4cab-9b94-e35003464890.png">

I'm new to MQTT, so I assume with MQTT explorer looking at the Mosquito MQTT server I see the topics from nixie-singlebody.  Under Home Assistant,

**Reply by rstrouse** [reply]
You have a space in the topic `nixie-single body` is not the same as `nixie-singlebody`

**Reply by Fumado** [reply]
Correct.  I can't modify that space since it comes from njsPC, but I don't know from which file.  I was able to edit the homeassistant.json to fix the spaces with the air and water temperature, but not the heater.

In any case, thanks for pointing me to the njsPC-HA, which has the heater info, so I think I'll try that for a while.

**Reply by tagyoureit** [reply]
There is a section in `./config.json` that has home assistant variables...
```
      "homeAssistant": {
        "name": "Home Assistant",
        "type": "mqtt",
        "enabled": false,
        "fileName": "homeassistant.json",
        "globals": {},
        "options": {
          "protocol": "mqtt://",
          "host": "192.168.0.1",
          "port": 1883,
          "username": "",
          "password": "",
          "selfSignedCertificate": false,
          "retain": true,
          "qos": 0,
          "changesOnly": true
        },
        "vars": {
          "_note": "hassTopic is the topic that HASS reads for configuration and should not be changed.  mqttTopic should match the topic in the MQTT binding (do not use MQTTAlt for HASS).",
          "hassTopic": "homeassistant",
          "mqttTopic": "@bind=(state.equipment.model).replace(' ','-').replace('/','').toLowerCase();"
        },
        "uuid": "fe583045-219e-47ac-8e34-55e7f6a29ccd"
      },
```

and in the `homeassistant.json` file the following line should pick up the system name and replace and spaces.  
```
"action_topic: `${ctx.vars.mqttTopic}/state/temps/bodies/${data.id}/${data.name.replace(/\\s+/g, '').toLowerCase()}/heatStatus`,",
```

edit: formatting
